### ISSUE TITLE
"relaxed parsing mode" for games? relaxed move rules...

### ISSUE THREAD

**Issue by davidschlangen** [clp-research/clemcore#27]
It looks like some of the open models fare worse than they perhaps have to, as they often seem to produce a good reply, but then keep on talking. E.g., they may do something like this for taboo:

```
Ok, that is a fun game!
GUESS: street
What do you think?
```

At the moment, our parsing rules are strict (in taboo, they want the output to be a single line prefixed with GUESS), and would count this as non-compliant. We can think about having more relaxed rules, that apply more of a "principle of charity" ("if something that looks like a well-formed reply is within the reply, extract it")... Maybe as a form of (optional) pre-processing on the reply?

The problem however is that this should not have to use any "intelligence"; i.e., it should be possible to formally define this preprocessing. Which might mean that we miss something that for a human would make the result invalid. (E.g., there being two different guesses in one output, or some other text that modifies the meaning of the guess, etc.)

**Comment by Gnurro** [comment]
I don't know how prompt pre-processing would help, since this is about parsing model outputs.
The currently 'too strict' parsing can be solved with very basic regEx matching.

**Comment by davidschlangen** [comment]
This was not at all an issue for you, but rather for the game designers. (But you're right, it should have said "pre-processing on the reply".)

**Comment by davidschlangen** [comment]
Bumping this up, as this is still relevant.

It should be possible to mostly do this via a) changing the regular expression used for parsing the response, and/or b) changing how it is applied (`fullmatch` vs `search`).

@phisad , is this (choice of regexp and matchin method) something that could be exposed by the framework? That is, make it a parameter on the outside call whether strict or liberal mode is used?

**Comment by phisad** [comment]
I like the example. So, as the prefix is defined by the game, all the framework can do is to offer a GM or a utility function that does this for the implementor. And I think the prefix should then live on the Player object. But anyways, the function could look like: `parse(response:str, indicator: str, strict:bool)` where indicator is e.g. `GUESS`. What could be possible cases?

Cases where we want to match `<indicator>:<text>\n`
```
Ok, that is a fun game!
GUESS: street
What do you think?
```
```
Ok, that is a fun game! GUESS: street
What do you think?
```
```
GUESS: street
```

Cases where we want to match `<indicator>:<text>.` (maybe also `!` or `?`)
```
Ok, that is a fun game! 
GUESS: street. What do you think?
```
```
Ok, that is a fun game! GUESS: street. What do you think?
```
```
GUESS: street.
```

We only want to return the first match, because they might hallucinate continuations like this:
```
Ok, that is a fun game!
GUESS: street
That's correct! Let's play another round. CLUE: "The animal that went over the street." Oh, that's a fun one. GUESS: chicken!
```

And especially the last one sets the question what exactly do we want to evaluate?

**Comment by davidschlangen** [comment]
Maybe it could be a cascade?
- first try *strict* (a correct reply contains only the response tag ("GUESS: ") and exactly as much material as is desired by the game; or, if the answers are categorial, exactly one member of the categoires);
-  if that fails, try *embellished* (find the first instance of tag + desired amount of information);
- if that fails (no tag present), try *hail mary* (extract desired amount of information, or first instance of a category)

**Comment by phisad** [comment]
Would we then have 3 different evaluations w.r.t the parsing strictness?

Just another issue came to my mind: Our approach does not allow for something like CoT-prompting where we would expect the last line of the chain to be the GUESS. 

Now, if the response continues even after that: would we simply cut-off everything after the guess or do we let the model fail. I tend to the latter. Similarly the _hail mary_ approach seems to mask the issue that the model does not follow the game rules set in the initial prompt.

**Comment by davidschlangen** [comment]
Yeah, no, just two. Strict and non-strict. The non-strict variant is only for getting an idea of how much a model loses by not being able to strictly follow the instructions. (I would expect the Hail Mary to not yield useful moves in most cases anyway.)

Good point about the CoT though. Maybe *strict* should be:
- a correct reply consists of TAG + correctly formed content, in order for each TAG that is expected. So for CoT, we'd expect "THOUGHT: free text as specified by regexp \n GUESS: free text as specified by regexp"

Oh dear. That gives me another idea. Maybe instead of (or in addition to) specifying the regexp for each game separately, we should have named regexps (like "free_text_one_par") that the game spec can reference, to give us some standardisation?

**Comment by AnneBeyer** [comment]
Another note: Ideally, it should be possible to change the scoring mode without re-running the benchmark, which is currently not the case because usually the game gets aborted if one player message fails to follow the strict mode. Maybe it would make sense to turn this around and have the liberal parsing mode as the default for running the benchmark, and then adding an option to apply the strict mode on top of the generated game plays in the scorer. (This would also ensure that both scoring modes are applied to the same content, avoiding the need to address the reproducibility issue here.)

**Comment by AnneBeyer** [comment]
And related to the idea of standardized regexes (which should maybe become a separate issue?): 
I'm not sure if this is applicable for all types of responses (I think some games even use regex-free answer parsing methods), but the multilingual referencegame version already uses two "skeleton" regexes (for free text and option selection) which then get filled with the language specific tags and options.

**Comment by phisad** [comment]
> Another note: Ideally, it should be possible to change the scoring mode without re-running the benchmark, which is currently not the case because usually the game gets aborted if one player message fails to follow the strict mode. Maybe it would make sense to turn this around and have the liberal parsing mode as the default for running the benchmark, and then adding an option to apply the strict mode on top of the generated game plays in the scorer. (This would also ensure that both scoring modes are applied to the same content, avoiding the need to address the reproducibility issue here.)

I rather like this idea. It would turn our evaluation more into a "recording" and an "annotation" part. But this is ok for now since the models do not learn between episodes of the game or between games anyway e.g. "Oh I did not follow the prefix rules in the episode before and thus lost the game, so now in episode 2 I should better care about the format!".

**Comment by davidschlangen** [comment]
Yeah, the idea has a certain charme. But you, @phisad , point at the problem: This would move us even further away from being a gym-like environment, when at the same time we have reasons to moving closer towards that. And ultimately, the liberal mode is just meant to give us an idea of how much of a factor the weakness in (form-related) instruction following is. For the real evaluation, I think "do exactly what I told you, or else I ignore you" is the cleanest principle.

**Comment by phisad** [comment]
@davidschlangen Not sure if we should further pursue this issue. What do you think? Anyways, the priority might be rather low.

**Comment by davidschlangen** [comment]
This is something that always comes up again, most recently in a discussion of the playpen evaluations. But I don't see a good principled decision that wouldn't overgenerate (i.e., accept something that shouldn't be accepted).
